src/dataframe/dataframe.jl

```diff
@@ -625,54 +617,78 @@ julia> insertcols!(d, 1, b=['a', 'b', 'c'])
-        if makeunique
-            k = 1
-            while true
-                # we only make sure that new column name is unique
-                # if df originally had duplicates in names we do not fix it
-                nn = Symbol("$(name)_$k")
-                if !hasproperty(df, nn)
-                    name = nn
-                    break
+    for (name, item) in name_cols
+        if ncol(df) == 0
+            if copycols
+                df[:, name] = item
+            else
+                df[!, name] = item
+            end
+            col_ind += 1
+            continue
+        end
+
+        if nrow(df) != length(item)
+            throw(DimensionMismatch("length of new column ($(length(item))) must match" *
+                                    " the number of rows in data frame ($(nrow(df)))"))
+        end
+
+        if hasproperty(df, name)
+            if makeunique
+                k = 1
+                while true
+                    nn = Symbol("$(name)_$k")
+                    if !hasproperty(df, nn)
+                        name = nn
+                        break
+                    end
+                    k += 1
                 end
-                k += 1
+            else
+                throw(ArgumentError("Duplicate variable name $name. Pass makeunique=true" *
+                                    " to make it unique using a suffix automatically."))
             end
+        end
+        insert!(index(df), col_ind, name)
+        if item isa AbstractRange
+            item_new = collect(item)
+        elseif copycols
+            item_new = copy(item)
         else
-            msg = "Duplicate variable name $name. Pass makeunique=true" *
-                  " to make it unique using a suffix automatically."
-            throw(ArgumentError(msg))
+            item_new = item
         end
+        insert!(_columns(df), col_ind, item_new)
+        col_ind += 1
     end
-    insert!(index(df), col_ind, name)
-    insert!(_columns(df), col_ind, item)
-    df
+    return df
 end
 
-insertcols!(df::DataFrame, col_ind::Int, name_col::Pair{Symbol}; makeunique::Bool=false) =
-    insertcols!(df, col_ind, name_col[1] => upgrade_scalar(df, name_col[2]), makeunique=makeunique)
+function insertcols!(df::DataFrame, col_ind::Int, name_col::Pair{Symbol}; makeunique::Bool=false)
+    Base.depwarn("Implicit broadcasting when inserting a column is deprecated. Pass an `AbstractVector` instead. ", :insertcols!)
+    # this is not a fully correct way to upgrade a scalar but this is how upgrade_scalar
+    # worked so I keep it not to be breaking the method will be removed anyways
+    insertcols!(df, col_ind, name_col[1] => fill(name_col[2], (ncol(df) == 0) ? 1 : nrow(df)), makeunique=makeunique)
+end
 
 function insertcols!(df::DataFrame, col_ind::Int; makeunique::Bool=false, name_col...)
+    Base.depwarn("inserting colums using a keyword argument is deprecated. Use `Pair` syntax instead. ", :insertcols!)
```

Comment:
Maybe the message could show what the new syntax looks like, e.g. `"insertcols!(df, col_ind, :$(keys(name_col)[1]) => v)`?

Reply:
ok